pkg/controller/common/reconciler.go

```
@@ -30,16 +31,37 @@ func (r *BaseIngressReconciler) ReconcileIngress(ctx context.Context, ci network
 
 	exposed := ci.GetSpec().Visibility == networkingv1alpha1.IngressVisibilityExternalIP
 	if exposed {
+		ingressLabels := ci.GetLabels()
+		listOpts := &client.ListOptions{
+			LabelSelector: labels.SelectorFromSet(map[string]string{
+				networking.IngressLabelKey:     ci.GetName(),
+				serving.RouteLabelKey:          ingressLabels[serving.RouteLabelKey],
+				serving.RouteNamespaceLabelKey: ingressLabels[serving.RouteNamespaceLabelKey],
+			}),
+		}
+		var existing routev1.RouteList
+		if err := r.Client.List(ctx, listOpts, &existing); err != nil {
+			return err
+		}
+		existingMap := routeMap(existing)
+
 		routes, err := resources.MakeRoutes(ci)
 		if err != nil {
 			return err
 		}
-
 		for _, route := range routes {
 			logger.Infof("Creating/Updating OpenShift Route for host %s", route.Spec.Host)
 			if err := r.reconcileRoute(ctx, ci, route); err != nil {
 				return err
 			}
+			delete(existingMap, route.Name)
+		}
+		// If routes remains in existingMap, it must be obsoleted routes. Clean them up.
+		for _, rt := range existingMap {
+			logger.Infof("Deleting obsoleted route: %s", rt.Name)
+			if err := r.deleteRoute(ctx, rt); err != nil {
+				logger.Warnf("Failed to delete obsoleted route %s: %v", rt.Name, err)
```

Why not actually return the error here? I guess the same goes for line 67.